File: Ruby/Convert_Hash/convert_hash.rb
# 入れ子の末端のキー名称「text」
KEY_TEXT = 'text'

def toUry(before_conversion_hash)
	before_conversion_hash.each do |key, value|
    # 入れ子の末端と合致する場合、「foo」を「uryyyy!!」に変換する。合致しない場合、現在の入れ子を引数にtoUryメソッドを呼び出す
    key == KEY_TEXT ? value.gsub!('foo', 'uryyyy!!') : toUry(value)
	end
end

# メイン処理
before_conversion_hash =
{
	'main' => {
		'first' => {'text' => 'foobar'},
		'second' => {'text' => 'fizzbuzz', 'child' => {'text' => 'foobar'}}
	},
	'sub' => {
		'first' => {'text' => 'fizzbuzz', 'child' => {'text' => 'foobar'}},
		'second' => {
			'third' => {'text' => 'barfoo', 'child' => {'text' => 'foobar'}},
			'forth' => {'child' => {'text' => 'jit_foo_foo'}}
		}
	},
	'text' => 'foofava'
}
puts toUry(before_conversion_hash)

File: TypeScript/Convert_Hash/convert_hash.ts
const KEY_TEXT = "text";

type JsonValue = string | number | boolean | null | JsonObject | JsonValue[];
type JsonObject = { [key: string]: JsonValue };

function toUry(beforeConversion: JsonObject): JsonObject {
  for (const [key, value] of Object.entries(beforeConversion)) {
    if (key === KEY_TEXT && typeof value === "string") {
      beforeConversion[key] = value.replace(/foo/g, "uryyyy!!");
      continue;
    }

    if (value && typeof value === "object" && !Array.isArray(value)) {
      toUry(value as JsonObject);
    }
  }
  return beforeConversion;
}

function main(): void {
  const beforeConversion: JsonObject = {
    main: {
      first: { text: "foobar" },
      second: { text: "fizzbuzz", child: { text: "foobar" } },
    },
    sub: {
      first: { text: "fizzbuzz", child: { text: "foobar" } },
      second: {
        third: { text: "barfoo", child: { text: "foobar" } },
        forth: { child: { text: "jit_foo_foo" } },
      },
    },
    text: "foofava",
  };

  const converted = toUry(beforeConversion);
  console.log(JSON.stringify(converted, null, 2));
}

main();

